File: app/payments/services/orchestrator.py
"""
Payment orchestration service for TerraFlow.

This module provides high-level orchestration of payment flows,
coordinating between different payment providers and cross-chain bridge operations.
"""

import logging
from typing import Dict, Any, Optional

from .payment_service import receive_fiat, send_fiat
from ..providers.bridge_adapter import get_bridge_provider

logger = logging.getLogger(__name__)

async def transfer_cross_border(
    user_id: str,
    amount: float,
    src_cc: str,
    dst_cc: str,
    currency: str,
    src_chain: str = "1", # Ethereum Mainnet chain ID
    dst_chain: str = "137", # Polygon chain ID
    recipient: str = None,
    metadata: Optional[Dict[str, Any]] = None
) -> Dict[str, Any]:
    """
    Execute a cross-border transfer using cross-chain bridges as an intermediary.
    
    This orchestrates a complete flow from source fiat currency to destination fiat currency,
    using Stargate Protocol as a bridge for the cross-border transfer.
    
    Flow:
    1. Pull funds from user's local payment source
    2. On-ramp via Stargate (unified liquidity pools across chains)
    3. Transfer across chains via Stargate (handled internally)
    4. Off-ramp to destination fiat
    5. Pay out to the recipient
    
    Args:
        user_id: ID of the user initiating the transfer
        amount: Amount to transfer
        src_cc: Source country code
        dst_cc: Destination country code
        currency: Currency to use (e.g., "USD")
        src_chain: Source blockchain chain ID (default: "1" for Ethereum)
        dst_chain: Destination blockchain chain ID (default: "137" for Polygon)
        recipient: Recipient wallet address on destination chain
        metadata: Optional additional data for the transaction
        
    Returns:
        A dictionary with the results of each step
        
    Raises:
        ValueError: If any part of the transfer fails
    """
    result = {
        "debit": None,
        "bridge_onramp": None, 
        "bridge_offramp": None,
        "payout": None,
        "status": "pending",
        "errors": []
    }
    
    try:
        # Step 1: Pull local fiat from user's payment source
        logger.info(f"Initiating fiat debit for user {user_id}, amount {amount} {currency}")
        result["debit"] = await receive_fiat(user_id, amount, currency)
        
        # Step 2: On-ramp via Stargate (unified liquidity pools across chains)
        logger.info(f"Bridging {amount} {currency} from chain {src_chain} to {dst_chain}")
        bridge = get_bridge_provider()
        result["bridge_onramp"] = await bridge.onramp(amount, currency, src_chain, dst_chain, recipient or user_id)
        
        # Step 3: Transfer across chains (handled internally by Stargate)
        logger.info(f"Cross-chain transfer complete with transaction ID: {result['bridge_onramp']['tx_id']}")
        
        # Step 4: Off-ramp to destination fiat
        logger.info(f"Off-ramping to fiat in destination country {dst_cc}")
        bank_account_id = metadata.get("bank_account_id") if metadata else None
        result["bridge_offramp"] = await bridge.offramp(amount, currency, dst_chain, bank_account_id or user_id)
        
        # Step 5: Pay out to the recipient
        logger.info(f"Sending fiat payout of {amount} {currency} to {dst_cc}")
        result["payout"] = await send_fiat(
            user_id=user_id, 
            amount=amount, 
            currency=currency, 
            country_code=dst_cc,
            metadata=metadata
        )
        
        result["status"] = "completed"
        return result
        
    except Exception as e:
        logger.error(f"Error during cross-border transfer: {str(e)}")
        result["status"] = "failed"
        result["errors"].append(str(e))
        
        # Attempt fallback and recovery based on where the failure occurred
        try:
            await _handle_transfer_fallback(result, user_id, amount, currency, src_cc)
        except Exception as fallback_error:
            logger.error(f"Fallback handling failed: {str(fallback_error)}")
            result["errors"].append(f"Fallback error: {str(fallback_error)}")
            
        return result

async def _handle_transfer_fallback(
    result: Dict[str, Any],
    user_id: str,
    amount: float,
    currency: str,
    country_code: str = "US"
) -> None:
    """
    Handle fallback logic for failed transfers.
    
    This attempts to recover from failures at different stages of the transfer process.
    
    Args:
        result: The current result dictionary with existing operation results
        user_id: User ID for the transfer
        amount: Amount being transferred
        currency: Currency being used
        country_code: Country code for refunds (default: US)
    """
    # Check where the failure occurred and apply appropriate fallback
    if result["debit"] and not result["bridge_onramp"]:
        # Failed after debit but before bridge on-ramp - refund the user
        logger.info(f"Initiating refund for user {user_id} of {amount} {currency}")
        await send_fiat(
            user_id=user_id, 
            amount=amount, 
            currency=currency, 
            country_code=country_code,
            refund=True
        )
        result["status"] = "refunded"
        
    elif result["bridge_onramp"] and not result["bridge_offramp"]:
        # Failed after on-ramp but before off-ramp
        # This is a complex state requiring manual intervention
        tx_id = result["bridge_onramp"].get("tx_id", "unknown") if isinstance(result["bridge_onramp"], dict) else getattr(result["bridge_onramp"], "tx_id", "unknown")
        logger.critical(
            f"Transfer in indeterminate state. On-ramped but not off-ramped. "
            f"Transaction ID: {tx_id}, Amount: {amount} {currency}"
        )
        result["status"] = "indeterminate_needs_review"
        
    elif result["bridge_offramp"] and not result["payout"]:
        # Failed after off-ramp but before payout
        # Need to retry the payout or notify operations team
        tx_id = result["bridge_offramp"].get("tx_id", "unknown") if isinstance(result["bridge_offramp"], dict) else getattr(result["bridge_offramp"], "tx_id", "unknown")
        logger.error(
            f"Funds off-ramped but payout failed. "
            f"Off-ramp ID: {tx_id}, Amount: {amount} {currency}"
        )
        result["status"] = "offramp_complete_payout_failed"
        
    # Add additional logic for other failure scenarios if needed 

File: app/payments/services/orchestrator.js
"use strict";
/**
 * Payment orchestration service for Liquicity.
 *
 * This module provides high-level orchestration of payment flows,
 * coordinating between different payment providers and cross-chain bridge operations.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferCrossBorder = transferCrossBorder;
const factory_1 = require("../providers/factory");
const payment_service_1 = require("./payment_service");
/**
 * Execute a cross-border transfer using cross-chain bridges as an intermediary.
 *
 * This orchestrates a complete flow from source fiat currency to destination fiat currency,
 * using Stargate Protocol as a bridge for the cross-border transfer.
 *
 * Flow:
 * 1. Pull funds from user's local payment source
 * 2. On-ramp via Stargate (unified liquidity pools across chains)
 * 3. Transfer across chains via Stargate (handled internally)
 * 4. Off-ramp to destination fiat
 * 5. Pay out to the recipient
 *
 * @param options Transfer options including amount, currencies, and chains
 * @returns A result object with the results of each step
 * @throws Error if any part of the transfer fails
 */
function transferCrossBorder(options) {
    return __awaiter(this, void 0, void 0, function* () {
        const { userId, amount, srcCountryCode, dstCountryCode, currency, srcChain = 1, // Default: Ethereum Mainnet
        dstChain = 137, // Default: Polygon
        recipient, metadata } = options;
        const result = {
            debit: null,
            bridgeOnramp: null,
            bridgeOfframp: null,
            payout: null,
            status: 'pending',
            errors: []
        };
        try {
            // Step 1: Pull local fiat from user's payment source
            console.log(`Initiating fiat debit for user ${userId}, amount ${amount} ${currency}`);
            result.debit = yield (0, payment_service_1.receiveFiat)(userId, amount, currency);
            // Step 2: On-ramp via Stargate (unified liquidity pools across chains)
            console.log(`Bridging ${amount} ${currency} from chain ${srcChain} to ${dstChain}`);
            const bridge = (0, factory_1.getBridgeProvider)();
            result.bridgeOnramp = yield bridge.onramp(amount, currency, srcChain, dstChain, recipient || userId);
            // Step 3: Transfer across chains (handled internally by Stargate)
            console.log(`Cross-chain transfer complete with transaction ID: ${result.bridgeOnramp.tx_id}`);
            // Step 4: Off-ramp to destination fiat
            console.log(`Off-ramping to fiat in destination country ${dstCountryCode}`);
            const bankAccountId = (metadata === null || metadata === void 0 ? void 0 : metadata.bankAccountId) || userId;
            result.bridgeOfframp = yield bridge.offramp(amount, currency, dstChain, bankAccountId);
            // Step 5: Pay out to the recipient
            console.log(`Sending fiat payout of ${amount} ${currency}`);
            result.payout = yield (0, payment_service_1.sendFiat)(userId, amount, currency);
            result.status = 'completed';
            return result;
        }
        catch (error) {
            console.error(`Error during cross-border transfer: ${error.message}`);
            result.status = 'failed';
            result.errors.push(error.message);
            // Attempt fallback and recovery based on where the failure occurred
            try {
                yield handleTransferFallback(result, userId, amount, currency);
            }
            catch (fallbackError) {
                console.error(`Fallback handling failed: ${fallbackError.message}`);
                result.errors.push(`Fallback error: ${fallbackError.message}`);
            }
            return result;
        }
    });
}
/**
 * Handle fallback logic for failed transfers.
 *
 * This attempts to recover from failures at different stages of the transfer process.
 *
 * @param result The current result object with existing operation results
 * @param userId User ID for the transfer
 * @param amount Amount being transferred
 * @param currency Currency being used
 */
function handleTransferFallback(result, userId, amount, currency) {
    return __awaiter(this, void 0, void 0, function* () {
        // Check where the failure occurred and apply appropriate fallback
        if (result.debit && !result.bridgeOnramp) {
            // Failed after debit but before bridge on-ramp - refund the user
            console.log(`Initiating refund for user ${userId} of ${amount} ${currency}`);
            yield (0, payment_service_1.sendFiat)(userId, amount, currency, true); // true = refund
            result.status = 'refunded';
        }
        else if (result.bridgeOnramp && !result.bridgeOfframp) {
            // Failed after on-ramp but before off-ramp
            // This is a complex state requiring manual intervention
            console.error(`Transfer in indeterminate state. On-ramped but not off-ramped. ` +
                `Transaction ID: ${result.bridgeOnramp.tx_id}, Amount: ${amount} ${currency}`);
            result.status = 'indeterminate_needs_review';
        }
        else if (result.bridgeOfframp && !result.payout) {
            // Failed after off-ramp but before payout
            // Need to retry the payout or notify operations team
            console.error(`Funds off-ramped but payout failed. ` +
                `Off-ramp ID: ${result.bridgeOfframp.tx_id}, Amount: ${amount} ${currency}`);
            result.status = 'offramp_complete_payout_failed';
        }
        // Add additional logic for other failure scenarios if needed
    });
}
